go/ephemeral/lib.go

```diff
@@ -92,25 +114,133 @@ func (e *EKLib) newDeviceEKNeeded(ctx context.Context, merkleRoot libkb.MerkleRo
 		return needed, err
 	}
 
-	return keybase1.Time(merkleRoot.Ctime())-ek.Metadata.Ctime >= KeyGenLifetimeSecs, nil
+	return keygenExpired(ek.Metadata.Ctime, merkleRoot), nil
+}
+
+func (e *EKLib) NewUserEKNeeded(ctx context.Context) (needed bool, err error) {
+	e.Lock()
+	defer e.Unlock()
+
+	merkleRootPtr, err := e.G().GetMerkleClient().FetchRootFromServer(ctx, libkb.EphemeralKeyMerkleFreshness)
+	if err != nil {
+		return false, err
+	}
+	return e.newUserEKNeeded(ctx, *merkleRootPtr)
 }
 
 func (e *EKLib) newUserEKNeeded(ctx context.Context, merkleRoot libkb.MerkleRoot) (needed bool, err error) {
+	defer e.G().CTrace(ctx, "newUserEKNeeded", func() error { return err })()
+
+	// Let's see what the latest server statement is.
+	statement, err := fetchUserEKStatement(ctx, e.G())
+	if err != nil {
+		return false, err
+	}
+	// No statement, so we need a userEK
+	if statement == nil {
+		return true, nil
+	}
+	// Can we access this generation? If not, let's regenerate.
 	s := e.G().GetUserEKBoxStorage()
-	maxGeneration, err := s.MaxGeneration(ctx)
+	ek, err := s.Get(ctx, statement.CurrentUserEkMetadata.Generation)
 	if err != nil {
-		return needed, err
+		switch err.(type) {
+		case *EKUnboxErr:
+			e.G().Log.Debug(err.Error())
+			return true, nil
+		default:
+			return false, err
+		}
 	}
-	if maxGeneration < 0 {
+	// Ok we can access the ek, check lifetime.
+	return keygenExpired(ek.Metadata.Ctime, merkleRoot), nil
+}
+
+func (e *EKLib) NewTeamEKNeeded(ctx context.Context, teamID keybase1.TeamID) (needed bool, err error) {
+	e.Lock()
+	defer e.Unlock()
+
+	merkleRootPtr, err := e.G().GetMerkleClient().FetchRootFromServer(ctx, libkb.EphemeralKeyMerkleFreshness)
+	if err != nil {
+		return false, err
+	}
+	statement, err := fetchTeamEKStatement(ctx, e.G(), teamID)
+	if err != nil {
+		return false, err
+	}
+	return e.newTeamEKNeeded(ctx, teamID, *merkleRootPtr, statement)
+}
+
+func (e *EKLib) newTeamEKNeeded(ctx context.Context, teamID keybase1.TeamID, merkleRoot libkb.MerkleRoot, statement *keybase1.TeamEkStatement) (needed bool, err error) {
+	// Let's see what the latest server statement is.
+	// No statement, so we need a teamEK
+	if statement == nil {
 		return true, nil
 	}
+	// Can we access this generation? If not, let's regenerate.
+	s := e.G().GetTeamEKBoxStorage()
+	ek, err := s.Get(ctx, teamID, statement.CurrentTeamEkMetadata.Generation)
+	if err != nil {
+		switch err.(type) {
+		case *EKUnboxErr:
+			e.G().Log.Debug(err.Error())
+			return true, nil
+		default:
+			return false, err
+		}
+	}
+	// Ok we can access the ek, check lifetime.
+	return keygenExpired(ek.Metadata.Ctime, merkleRoot), nil
+}
 
-	ek, err := s.Get(ctx, maxGeneration)
+func (e *EKLib) GetOrCreateLatestTeamEK(ctx context.Context, teamID keybase1.TeamID) (teamEK keybase1.TeamEk, err error) {
+	// TODO put an LRU cache in front of this with a short lifetime so we don't
+	// hit the server ever time we want the latest key for a team.
```

Comment:
In the same vein, I wonder if picking 30 seconds for EphemeralKeyMerkleFreshness was too aggressive. That one's also a round trip. Not sure yet.

Reply:
Hmm good to keep in mind.